// @hyext/hyext-rn-sdk/views/HYRNVideoAnimationView/index.android.js
import React, { Component } from 'react';
import { View } from 'react-native';
import PropTypes from 'prop-types';

import hyRequireNativeComponent from '../hyRequireNativeComponent';

const NativeVideoAnimationView = hyRequireNativeComponent('HYRNAlphaVideoView');

export default class HYRNVideoAnimation extends Component {

    static props = {
        ...View.propTypes,
        src: PropTypes.string,
        loop: PropTypes.bool,
        onStart: PropTypes.func,
        onEnd: PropTypes.func,
        onError: PropTypes.func
    }

    render() {
        return (
            <NativeVideoAnimationView
                {...this.props}
                ref={(ref) => this.viewRef = ref}
                src={this.props.src || null}
                loop={this.props.loop || false}
                onStart={() => {
                    this.props.onStart && this.props.onStart();
                }}
                onEnd={() => {
                    this.props.onEnd && this.props.onEnd();
                }}
                onError={(err) => {
                    this.props.onError && this.props.onError(err);
                }}
            />
        );
    }
}
