// @hyext/cli/constants.js
module.exports = {
  buildTypesMap: {
    '主站面板': {
      extType: 'web_video_com',
      platform: 'web',
    },
    '主站弹窗': {
      extType: 'web_popup',
      platform: 'web',
    },
    'PC 客户端主播面板': {
      extType: 'pc_anchor_panel',
      platform: 'pc',
    },
    'PC 客户端观众面板': {
      extType: 'pc_panel',
      platform: 'pc',
    },
    'PC 客户端观众弹窗': {
      extType: 'pc_popup',
      platform: 'pc',
    },
    '虎牙直播 APP 面板': {
      extType: 'app_panel',
      platform: 'app',
    },
    '虎牙直播 APP 弹窗': {
      extType: 'app_popup',
      platform: 'app',
    },
    '虎牙助手 APP 面板': {
      extType: 'zs_anchor_panel',
      platform: 'app',
    },
    '虎牙助手 APP 弹窗': {
      extType: 'zs_anchor_popup',
      platform: 'app',
    },
  },
  extTypes: [
    {
      extType: 'web_video_com',
      platform: 'web',
      longName: '主站面板',
      name: '面板',
      role: 'audience',
    },
    {
      extType: 'web_popup',
      platform: 'web',
      longName: '主站浮窗',
      name: '浮窗',
      role: 'audience',
    },
    {
      extType: 'pc_anchor_panel',
      platform: 'pc',
      longName: 'PC 主播端面板',
      name: '面板',
      role: 'streamer',
    },
    {
      extType: 'pc_panel',
      platform: 'pc',
      longName: 'PC 观众端面板',
      name: '面板',
      role: 'audience',
    },
    {
      extType: 'pc_popup',
      platform: 'pc',
      longName: 'PC 观众端浮窗',
      name: '浮窗',
      role: 'audience',
    },
    {
      extType: 'app_panel',
      platform: 'app',
      longName: '虎牙 APP 面板',
      name: '面板',
      role: 'audience',
    },
    {
      extType: 'app_popup',
      platform: 'app',
      longName: '虎牙 APP 浮窗',
      name: '浮窗',
      role: 'audience',
    },
    {
      extType: 'zs_anchor_panel',
      platform: 'app',
      longName: '虎牙助手面板',
      name: '面板',
      role: 'streamer',
    },
    {
      extType: 'zs_anchor_popup',
      platform: 'app',
      longName: '虎牙助手浮窗',
      name: '浮窗',
      role: 'streamer',
    },
  ]
}
